### PR TITLE
Round from zero support for non-BigFloats

### PR DESCRIPTION
**Pull request by jessymilare**
My attempt to fix both #34519 and #41182.

Implements `round(x, RoundFromZero)`, `div(x, y, RoundFromZero)` and friends.

### CODE REVIEW

File: base/div.jl [hunk 1]
```diff
@@ -83,6 +91,10 @@ rem(x, y, ::RoundingMode{:Up}) = mod(x, -y)
 rem(x, y, r::RoundingMode{:Nearest}) = x - y*div(x, y, r)
 rem(x::Integer, y::Integer, r::RoundingMode{:Nearest}) = divrem(x, y, r)[2]
 
+function rem(x, y, ::typeof(RoundFromZero))
+    rem(x, y, ifelse(signbit(x) == signbit(y), RoundUp, RoundDown))
```
Comment (mcabbott) [inline base/div.jl:95]: I think this is a type instability, might be better to branch outside?
```
signbit(x) == signbit(y) ? rem(x, y, RoundUp) : rem(x, y, RoundDown)
```
Although when I tried to benchmark that, I didn't see a timing difference between the two.

Comment (jessymilare) [reply base/div.jl:95]: I tested with `BigInt`s and I also see no difference in performance (compiled Julia with `JULIA_CPU_TARGET=native` and `MARCH=native`).

I tend to think you are right, though. The compiler seems to be able to handle that type instability, but there might be some corner cases in which it doesn't.

**Comment by StefanKarpinski** [comment]
This appears complete. Who needs to review this?

**Review by oscardssmith — APPROVED** [review]
Looks good! I'm pretty sure there is a more efficient way to do this rounding, but it's fine to add this with a less than optimal implementation, and improve it later if anyone cares enough.

File: base/div.jl [hunk 2]
```diff
@@ -30,6 +30,10 @@ julia> div(-5, 2, RoundNearestTiesAway)
 -3
 julia> div(-5, 2, RoundNearestTiesUp)
 -2
+julia> div(4, 3, RoundFromZero)
```
Comment (mcabbott) [inline base/div.jl:33]: Should there be a "!! compat needs Julia 1.8" message somewhere? Perhaps in this docstring for `round`, perhaps elsewhere?

**Comment by StefanKarpinski** [comment]
Have restarted CI since June 16th was a while ago.

File: base/div.jl (hunk 2 above)
Comment (jessymilare) [reply base/div.jl:36]: Added notes to docs in `RoundFromZero`, `div`, `rem` and `RoundingMode`.

**Comment by jessymilare** [comment]
Only failing on FreeBSD. Log only show a bunch of EOFErrors.

**Comment by stevengj** [comment]
Fixed conflicts; hopefully this is good to go?

File: base/div.jl [hunk 3]
```diff
@@ -17,6 +17,9 @@ without any intermediate rounding.
 
 See also [`fld`](@ref) and [`cld`](@ref), which are special cases of this function.
 
+!!! compat "Julia 1.8"
+    `RoundFromZero` requires at least Julia 1.8.
```
Comment (stevengj) [inline base/div.jl:21]:
```suggestion
    `RoundFromZero` requires at least Julia 1.9.
```

File: base/div.jl [hunk 4]
```diff
@@ -63,6 +70,13 @@ without any intermediate rounding.
   `[0,-y)` otherwise. The result may not be exact if `x` and `y` have the same sign, and
   `abs(x) < abs(y)`. See also [`RoundUp`](@ref).
 
+- if `r == RoundFromZero`, then the result is in the interval `(-y, 0]` if `y` is positive, or
+  `[0, -y)` otherwise. The result may not be exact if `x` and `y` have the same sign, and
+  `abs(x) < abs(y)`. See also [`RoundFromZero`](@ref).
+
+!!! compat "Julia 1.8"
+    `RoundFromZero` requires at least Julia 1.8.
```
Comment (stevengj) [inline base/div.jl:78]:
```suggestion
    `RoundFromZero` requires at least Julia 1.9.
```

File: base/rounding.jl [hunk 5]
```diff
@@ -37,9 +37,13 @@ Currently supported rounding modes are:
 - [`RoundNearestTiesAway`](@ref)
 - [`RoundNearestTiesUp`](@ref)
 - [`RoundToZero`](@ref)
-- [`RoundFromZero`](@ref) ([`BigFloat`](@ref) only)
+- [`RoundFromZero`](@ref)
 - [`RoundUp`](@ref)
 - [`RoundDown`](@ref)
+
+!!! compat "Julia 1.8"
+    `RoundFromZero` requires at least Julia 1.8. Prior versions support
```
Comment (stevengj) [inline base/rounding.jl:45]:
```suggestion
!!! compat "Julia 1.9"
    `RoundFromZero` requires at least Julia 1.9. Prior versions support
```

File: base/rounding.jl [hunk 6]
```diff
@@ -76,15 +80,18 @@ const RoundDown = RoundingMode{:Down}()
     RoundFromZero
 
 Rounds away from zero.
-This rounding mode may only be used with `T == BigFloat` inputs to [`round`](@ref).
+
+!!! compat "Julia 1.8"
+    `RoundFromZero` requires at least Julia 1.8. Prior versions support
```
Comment (stevengj) [inline base/rounding.jl:85]:
```suggestion
!!! compat "Julia 1.9"
    `RoundFromZero` requires at least Julia 1.9. Prior versions support
```

**Review by stevengj — APPROVED** [review]

File: NEWS.md [hunk 7]
```diff
@@ -41,6 +41,7 @@ Library changes
   as `keys(::Dict)`, `values(::Dict)`, and `Set` is fixed.  These methods of `iterate` can
   now be called on a dictionary or set shared by arbitrary tasks provided that there are no
   tasks mutating the dictionary or set ([#44534]).
+* `RoundFromZero` now works for non-`BigFloat` types ([#41246]).
```
Comment (giordano) [inline NEWS.md:44]: Trailing whitespace
```suggestion
* `RoundFromZero` now works for non-`BigFloat` types ([#41246]).
```

**Pull request merged**